### PR TITLE
Add tests for various T::Props edge cases and less common APIs

### PR DESCRIPTION
**Pull request by djudd-stripe**
Newly tested:
- `validate_prop_value`
- T.untyped as a prop type (accessors)
- Custom types as prop types (accessors, serialization)
- T::Enum as prop type (serialization)
- Defaults & factories when deserializing to required prop
- Support for overriding `prop_get` on a decorator subclass
- Serialization under weird inheritance scenarios
- Serialization with empty struct (no props)
- Behaviors of setters & constructors with `raise_on_nil_write` and with/without explicit default

### Motivation
Test scenarios that broke without failing tests in various optimization experiments

Setup for actually PR-ing some of those experiments, same as https://github.com/sorbet/sorbet/pull/2401

### Test plan
This only adds tests

### CODE REVIEW

File: gems/sorbet-runtime/test/types/configuration.rb [hunk 1]
```diff
@@ -198,15 +198,15 @@ def @mod.foo(a)
     describe 'scalar_types' do
       describe 'when overridden' do
         before do
-          T::Configuration.scalar_types = ['foo']
+          T::Configuration.scalar_types = T::Configuration.scalar_types.dup << 'foo'
```
Comment (djudd-stripe) [inline gems/sorbet-runtime/test/types/configuration.rb]: This is a workaround for the fact that this mutates a global variable, making it racy with other tests. (The race is still there in some sense, just doesn't cause other tests to fail)

Comment (DarkDimius) [reply gems/sorbet-runtime/test/types/configuration.rb]: Could you give more context here? Why is it racy and can we make it not be racy?

Comment (djudd-stripe) [reply gems/sorbet-runtime/test/types/configuration.rb]: We're setting a global variable here that we check in other tests, so any that read it (which anything using custom types does IIRC) will be affected if they run simultaneously. I'm not actually sure why other tests were running in parallel (or interspersed somehow?) with this one but empirically it happened.
